Search: Add pricing and tier information to Plans page

jsnmoon opened the pull request:
Jetpack equivalent to Automattic/wp-calypso#40306.

#### Changes proposed in this Pull Request:
* Adds support for `/sites/<id>/products` API route via `<QuerySiteProducts />`, `site-products` state handling, and `jetpack/v4/site/products` local API route.
* Adds handling for showing Search plan tiers
* Adds handling for showing Search plan prices

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Yes, this adds support for showing Search plan tiers to the wp-admin Jetpack plans page. This is being merged into a feature branch, however.

#### Testing instructions:
1. You must apply Phabricator diff 153244 to your sandbox. This is a variant of D38944-code.

2. Apply changes from this PR to your Jetpack installation. If you're running a dockerized instance, you may need to [update your `extra_hosts`](https://docs.docker.com/compose/compose-file/#extra_hosts) to properly sandbox the WPCOM API.

3. Navigate to the Jetpack Plans page at `/wp-admin/admin.php?page=jetpack#/plans`. 

4. Ensure that both the record counts and the tier information render as expected.

<img width="1068" alt="Screen Shot 2020-03-25 at 6 10 37 PM" src="https://user-images.githubusercontent.com/4044428/77597305-0bde8800-6ec4-11ea-9fe6-8978890572e7.png">

#### Proposed changelog entry for your changes:
* None for now.

Commit:
Add handling for pricing and tier APIs

jsnmoon commented:
@keoshi: Due to the slower response times for the Tiers API, I split out the loading states for Backups and Search. Unfortunately, that means this is now possible:

<img width="1068" alt="Screen Shot 2020-03-25 at 4 27 43 PM" src="https://user-images.githubusercontent.com/4044428/77591492-abe0e500-6eb5-11ea-9f30-7eca7a053f8f.png">

Should I unify the loading states for Backups and Search so that we could avoid this misalignment?

gibrown commented on _inc/client/plans/single-product-search/index.jsx:
```diff
@@ -2,21 +2,53 @@
+function getTierLabel( priceTierSlug, recordCount ) {
+	switch ( priceTierSlug ) {
+		case JETPACK_SEARCH_TIER_UP_TO_100_RECORDS:
+			return __( 'Tier 1: Up to 100 records' );
```
Let's remove the numbering for the tiers here and just say the number of records. We may add new tiers at some point.

jsnmoon replied:
👍

gibrown commented on _inc/client/plans/single-product-search/index.jsx:
```diff
@@ -37,20 +69,23 @@ export default function SingleProductSearchCard( props ) {
 					) }
 				</h4>
 				<div className="single-product-search__radio-buttons-container">
 					<PlanRadioButton
```
@keoshi can we get a different design for what we want to do here. The radio button feels weird unless we are going to have the selection be between monthly and yearly.

For yearly we could then say "two months free" or something and show them side by side.

jsnmoon replied:
This has since been updated, but I think we'd benefit greatly from some design direction on this part of the interface.

keoshi replied:
Yeah, this is still a long way from an optimal solution, but since it has to be articulated with the Backup team, I'd say let's leave it for now. I will start that conversation with the design team as soon as possible having this mock as a starting point:

![image](https://user-images.githubusercontent.com/390760/77677861-3560e780-6f88-11ea-89d6-2056da32319b.png)

gibrown reviewed (commented):
Minor copy/design comments...

Commit:
Remove tier numbers

Commit:
Add monthly and yearly pricing into search plan

Commit:
Add support for showing ProductSavings component

jsnmoon commented:
- Updated to include monthly and yearly pricing options. The monthly upgrade button doesn't redirect correctly at the moment.
- Added a snippet about how much the customer is saving or would save using the yearly pricing option.

DavidRothstein commented on _inc/client/plans/single-product-search/index.jsx:
```diff
@@ -1,22 +1,60 @@
+	const handleSelectedTimeframeChange = handleSelectedTimeframeChangeFactory( setTimeframe );
+	const currencyCode = get( props.products, 'jetpack_search.currency_code', '' );
+	const monthlyPrice = get( props.products, 'jetpack_search_monthly.cost', '' );
+	const yearlyPrice = get( props.products, 'jetpack_search.cost', '' );
```
I think you need to get all these properties from `props.siteProducts` rather than `props.products` (in fact, `props.products` can probably be removed from this component entirely)?

`siteProducts` is the one that has the site-specific pricing (based on the current search tier) so without that I think the wrong price will be shown.

jsnmoon replied:
Gotcha, I'll make the fix.

DavidRothstein reviewed (commented):
Looks pretty good to me (with the caveat that I didn't actually test it).  However, I think there's one thing that may be broken -- see inline comment.

keoshi commented:
@jsnmoon 

> Should I unify the loading states for Backups and Search so that we could avoid this misalignment?

I don't if there's prior art in doing something like this or a best practice in terms of performance, but I'd personally much rather see them show up at the same time, yes! :)

gibrown commented:
This is looking good to me, I think we are waiting for the wp.com deploy anyways, so will hold off on a final review until that lands.

Commit:
Use site products for pricing and currency code

Commit:
Use a unified loading state for single feature products

DavidRothstein commented on _inc/client/plans/product-selector.jsx:
```diff
@@ -256,11 +256,14 @@ export default connect( state => {
 		multisite: isMultisite( state ),
 		products: getProducts( state ),
```
I think in one of the earlier comments you mentioned this component got slower to load (due to the extra API endpoint)...

I think you can actually eliminate this `products` property from the component (and the `isFetchingProducts` property below) and just pass `siteProducts` in to `SingleProductBackup` instead -- this should work fine since the two have the same structure (and for Jetpack Backup, the exact same data).

In that case, you can probably also avoid calling `QueryProducts` on this page entirely, which would avoid having to hit that API endpoint at all.

gibrown reviewed (approved):
Tested this along with D40938 and it seems to all work well. The redirects are currently not working because we need to move the search plans to staging, but they all look correct based on inspection.
